source/Cosmos.HAL2/Drivers/PCI/Audio/AC97.cs

```diff
@@ -0,0 +1,301 @@
+        /// <param name="mute">Whether to mute all output of the channel.</param>
+        private static ushort CreateMixerVolumeValue(byte right, byte left, bool mute)
+            => (ushort)((right & 0x3f) | ((left & 0x3f) << 8) | ((mute ? 1 : 0) & 1 << 15));
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="AC97"/> class, with the
+        /// given buffer size.
+        /// </summary>
+        /// <param name="bufferSize">The buffer size in samples to use. This value cannot be an odd number, as per the AC97 specification.</param>
+        /// <exception cref="ArgumentException">Thrown when the given buffer size is incorrect.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when no AC97-compatible sound card is present.</exception>
+        private AC97(ushort bufferSize)
+        {
+            if (bufferSize % 2 != 0)
+                // As per the AC97 specification, the buffer size cannot be odd.
+                // (1.2.4.2 PCM Buffer Restrictions, Intel document 302349-003)
+                throw new ArgumentException("The buffer size must be an even number.", nameof(bufferSize));
+
+            PCIDevice pci = Cosmos.HAL.PCI.GetDeviceClass(
+                ClassID.MultimediaDevice, // 0x04
+                (SubclassID)0x01          // 0x01
+            );
+
+            if (pci == null || !pci.DeviceExists || pci.InterruptLine > 0xF)
+                throw new InvalidOperationException("No AC97-compatible device could be found.");
+
+            PCI = pci; // Expose PCI device to the public API
+
+            pci.EnableBusMaster(true);
+            pci.EnableMemory(true);
+            pci.EnableDevice(); // enable I/O space
+
+            INTs.SetIrqHandler(pci.InterruptLine, HandleInterrupt);
+
+            ushort NAMbar = (ushort)pci.BaseAddressBar[0].BaseAddress;  // Native Audio Mixer
+            ushort NABMbar = (ushort)pci.BaseAddressBar[1].BaseAddress; // Native Audio Bus Master
+
+            pTransferControl = new IOPort((ushort)(NABMbar + 0x1B));
+            pMasterVolume = new IOPort((ushort)(NAMbar + 0x02));
+            pPCMOutVolume = new IOPort((ushort)(NAMbar + 0x18));
+            pBufferDescriptors = new IOPort((ushort)(NABMbar + 0x10));
+            pTransferStatus = new IOPort((ushort)(NABMbar + 0x16));
+            pLastValidEntry = new IOPort((ushort)(NABMbar + 0x15));
+            pGlobalControl = new IOPort((ushort)(NABMbar + 0x2C));
+            pResetRegister = new IOPort((ushort)(NAMbar + 0x00));
+
+            // Reset device
+            pGlobalControl.Byte = 0x2;
+            pResetRegister.DWord = 0xDEADBEEF; // any value will do here
+
+            // Reset PCM out
+            pTransferControl.Byte = (byte)(pTransferControl.Byte | TC_TRANSFER_RESET);
+            while ((pTransferControl.Byte & TC_TRANSFER_RESET) != 0)
+            {
+                // Wait until the byte is cleared
```

Comment:
Will this always occur? Otherwise can you add some timeout.

Reply:
Yep, this won't occur on cards such as the Ensoniq AudioPCI. I'm working on a driver for it and the timeout is added in my local version, the driver should be out soon :3